Use route names

kalilsn opened the pull request:
https://github.com/department-of-veterans-affairs/vets-contrib/issues/3287

Switches us over to use route names as the primary identifier.

Something I learned during this work is that ALL kong entities have a `PUT` endpoint which will upsert the entity. For routes, services, consumers, upstreams, and SNIs, either a UUID or a name can be used in this endpoint. If a UUID is specified, kong will use that UUID. This means that instead of altering kongfig to use names as the primary key for routes, we could instead allow it to use both, and require a UUID to be defined in the config file. Instead of kong generating the UUID, we would do that ourselves. That would prevent us from needing to write back updates from kong to our config files. It would also allow us to rename routes (and services, etc. if we adopted that approach for them as well).

Commit:
Include route names in kong api requests

Since kong 1.0 introduced route names, kongfig's custom route name
handling doesn't play nicely with the official attribute. This commit
lets kongfig use the new route names while still allowing the old format
to work.

Commit:
Ensure attributes.name is always set

Commit:
Add some tests for route name changes

Commit:
Revert "Ensure attributes.name is always set"

This reverts commit b58511dfbedbe38c3ae8d3ff3597072e10b05603.

Commit:
Remove useless tests

Commit:
Use route names as primary identifier

Commit:
Update route integration test

These tests (updating route names) no longer make sense: now that names
are the primary identifier, kongfig can't update them. For the same
reason, a route without a name (the previous starting point of the
tests) is no longer a valid state for kongfig.

dvogel commented:
I'm open to either direction. I think the most important goal is having a consistent user experience across services, routes, etc. I'm intrigued by the possibility of renaming routes (and other entities) but I'm also worried that allowing UUIDs to be specified manually will lead to copy/paste changes to cause us to duplicate UUIDs across environments. That wouldn't technically cause any issues with kong but it could lead to some real consternation and confusion. Our config linter could help with that. Perhaps we should move ahead with our plan to remove route IDs and then reconsider this potential new approach after we complete our kong declarative config spike.

kalilsn commented:
>I'm open to either direction. I think the most important goal is having a consistent user experience across services, routes, etc. I'm intrigued by the possibility of renaming routes (and other entities) but I'm also worried that allowing UUIDs to be specified manually will lead to copy/paste changes to cause us to duplicate UUIDs across environments. That wouldn't technically cause any issues with kong but it could lead to some real consternation and confusion. Our config linter could help with that. Perhaps we should move ahead with our plan to remove route IDs and then reconsider this potential new approach after we complete our kong declarative config spike.

I agree with all of this. If we go ahead with the alternate plan, I think we'd want to make a few tooling changes to go along with the new process. Definitely just going to finish this up for now, but let's revisit after the spike!

jasalisbury reviewed (approved):
:+1:

Commit:
Merge branch 'master' into kalilsn/use-route-names

Commit:
Error if routes don't have names

Commit:
Add better route test

Commit:
Allow both locations of route name in validation

Commit:
Test that error is thrown when route has no name

Commit:
Add more tests and stop supporting attributes.name for routes

This matches how kongfig treats other entities (like services) that
have a name property which can be used as a primary key. Supporting both
makes the code overly complicated and is inconsistent.

Commit:
Fix missing name test

Commit:
Remove stray console log

Commit:
Remove unneeded name to id translation

Commit:
Update snapshots

There shouldn't have been a change to the snapshot from this work, but
there was because the snapshots contained a bug. Previously the route
was being recreated (because it had no id) during the tests – now that
request is accurately snapshotted as a no-op.

Commit:
Merge branch 'master' into kalilsn/use-route-names

kalilsn commented on src/cli-apply.js:
```diff
@@ -78,35 +78,11 @@ else {
-        let updatedConfig = updateRouteIds(config, out);
-        updatedConfig = sanitizeConfigForSafeWrite(updatedConfig, envVarPointers);
+    .then(state => {
+        config = sanitizeConfigForSafeWrite(config, envVarPointers);
```
Now that we don't update route ids, it's clearer that there's really no changes from kongfig apply that need to be written back to the file. I'm not deleting this in case we decide we do want that to be true at some point in the future

maddy-jo replied:
this is ironic to me because we are now going from the `if` block being executed exclusively to the `else` block being executed exclusively

kalilsn commented on src/kongState.js:
```diff
@@ -34,9 +34,9 @@ export default async (adminApi) => {
         servicesWithPluginsAndRoutes = await Promise.all(services.map(async item => {
             const plugins = await adminApi.fetchServicePlugins(item.id);
-            const routes = await adminApi.fetchServiceRoutes(item.name);
+            const routes = await adminApi.fetchServiceRoutes(item.id);
```
This was an unintentional use of the service name instead of id. It worked fine because kong accepts either as a primary id. Changing this anyways, since it was a red herring that tripped me up during some debugging.

Commit:
Remove logging statements

kalilsn commented on src/reducers/services.js:
```diff
@@ -6,15 +6,17 @@ const plugins = (state, log, version) => {
 
   case 'add-route-plugin':
-  case 'add-service-plugin': return [ ...state, parsePlugin(content, version) ];
+  case 'add-service-plugin':
```
Lots of whitespace changes in this file, purely to help me read it better. We should probably consistently format this whole codebase soon.

jasalisbury replied:
Once we have this built in a CI server we can add a linter to enforce formatting.

kalilsn commented on src/reducers/services.js:
```diff
@@ -26,33 +28,33 @@ const routes = (state, log, version) => {
+    case 'add-service-route':
+      return [
+        ...state.filter(route => route.name !== params.routeName),
+        {...parseRoute(content), plugins: [] }
```
This should be equivalent, since an add-service-route action will always include the route name. Previously the name would have actually been an `id`.

kalilsn commented on test-integration/routes.test.js:
```diff
@@ -1,89 +1,97 @@
+        name: "foo",
+        paths: expect.arrayContaining(["/foo", "/bar"]),
+        regex_priority: 3
+    });
```
These integration tests are a little different from the ones we already have. Rather than snapshot testing the logs, I'm just asserting various things about the state of kong, which gives me more confidence in these changes.

kalilsn replied:
If this approach works for us, we'll probably end up wanting some utilities to make these tests easier to write.

Commit:
3.0.0

kalilsn commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@adhocteam/kongfig",
-  "version": "2.2.0",
+  "version": "3.0.0",
```
I gave this a major version because it's a breaking change

maddy-jo commented on src/readKongApi.js:
```diff
@@ -100,16 +100,8 @@ function parseServiceOrRoutePlugins(plugins, version) {
-        }
-    }
-    return { name: (name || id), id, attributes: {...rest}, _info: { id, updated_at, created_at } };
+export function parseRoute({ name, id, created_at, updated_at, service, ...attributes }) {
```
```suggestion
export function parseRoute({ name, id, created_at, updated_at, ...attributes }) {
```

unused?

kalilsn replied:
I think the idea is to explicitly remove `service` from the route. I don't exactly know why, but it seems to work right now

jasalisbury replied:
Yea we actually just took a look at this. Removing service from the params causes it to be added to attributes and show up on the route that gets returned. I would _think_ Kong can handle having the service field on routes but do you think we should keep this as is for now?

kalilsn replied:
My guess is that it's because kongfig expresses the relationship between services and routes as routes being children of services. So any time we're working with routes its in the context of a service to start (and therefore don't need the service reference in the object).

Technically kong allows you to create a route without a service, and I suspect kongfig wouldn't support that. At some point in the future we may want to rework that.

kalilsn replied:
@jasalisbury does that cause any errors? Do the integration tests pass still? I'm really fine either way, just wanted to point out that it's not actually a no-op (although it is subtle).

jasalisbury replied:
@kalilsn It does cause a snapshot error. I'm inclined to leave the param as is to not change too much at once. Thoughts @kalilsn and @mjlumetta?

maddy-jo replied:
sound logic, i support that

jasalisbury committed:
Merge branch 'master' into kalilsn/use-route-names
